sdk/core/core-rest-pipeline/src/retryStrategies/failoverRetryStrategy.ts

```diff
@@ -0,0 +1,173 @@
+    const maxRetryDelayInMs = factoryOptions.maxRetryDelayInMs ?? DEFAULT_CLIENT_MAX_RETRY_INTERVAL;
+
+    while (true) {
+      const request = retryState.response?.request || retryState.responseError?.request;
+      if (!request) {
+        throw Error("Request should be defined.");
+      }
+
+      // Select correct list of hosts for this request
+      const isSafeMethod = ["GET", "HEAD", "OPTIONS"].includes(request.method);
+      const currentStateList = isSafeMethod ? readHostStateList : writeHostStateList;
+
+      if (currentStateList.hostStates.length === 0) {
+        retryState = yield undefined;
+        continue;
+      }
+      const hostState = currentStateList.hostStates[currentStateList.index];
+
+      const requestFireTime = Math.max(hostState.retryAt?.valueOf() ?? -Infinity, Date.now());
+
+      retryState = yield hostState;
+
+      const exponentialDelay = _retryAfterUtil.exponentialDelayInMs(
+        hostState.retryCount,
+        retryDelayInMs,
+        maxRetryDelayInMs
+      );
+      const nextRetryAt = new Date(requestFireTime + exponentialDelay);
+      hostState.retryAt = nextRetryAt;
+
+      currentStateList.index = (currentStateList.index + 1) % currentStateList.hostStates.length;
+      hostState.retryCount++;
+    }
+  };
+}
+
+/**
+ *
+ * @param options - Lists of fallback endpoints for read/write operations
+ * @returns An exponential retry strategy with support for multiple endpoints
+ */
+export function failoverRetryStrategy(options: DefaultRetryPolicyOptions): RetryStrategy {
+  const delegateMap = new WeakMap<PipelineRequest, ReturnType<FailoverHostDelegate>>();
+  return {
+    name: "failoverRetryStrategy",
+    retry(state) {
+      if (!shouldFailover(state) || !options.failoverHostDelegate) {
+        return { skipStrategy: true };
+      }
+
+      const request = state.response?.request || state.responseError?.request;
+      if (!request) {
+        throw Error("Request should be defined.");
+      }
+
+      // Register a delegate for failover hosts for this request (and all of its retries)
+      if (!delegateMap.has(request)) {
+        delegateMap.set(request, options.failoverHostDelegate(state, options));
+      }
+
+      // Get the next host for this request
+      const failoverHosts = delegateMap.get(request)!;
+      const hostResult = failoverHosts.next(state);
+
+      // Skip if there are no fallback hosts
+      if (hostResult.done) {
+        return { skipStrategy: true };
+      }
+
+      // Skip if the delegate decides not to failover
+      const hostState = hostResult.value;
+      if (!hostState) {
+        return { skipStrategy: true };
+      }
+
+      const redirectUrl = new URL(request.url);
+      redirectUrl.host = hostState.host;
+
+      const modifiers: RetryModifiers = { redirectTo: redirectUrl.toString() };
+
+      if (hostState.retryAt) {
+        const timeUntilRetryAt = hostState.retryAt.valueOf() - Date.now();
```

Comment:
nit: this is functionally equivalent, but I think a little more readable. It's rare we call `valueOf` directly, and I think `getTime` makes it a bit more obvious that we are converting to ticks.

```suggestion
        const timeUntilRetryAt = hostState.retryAt.getTime() - Date.now();
```

Reply:
That makes sense. Fixed!